include/irr/asset/material_compiler/IR.h

```diff
@@ -0,0 +1,429 @@
+            { 
+                auto found = std::find_if(begin(),end(),[s](const INode* child){return child->symbol==s;});
+                if (found != (array+count))
+                {
+                    if (ix)
+                        ix[0] = found-array;
+                    return true;
+                }
+                return false;
+            }
+
+            operator bool() const { return count!=0ull; }
+
+            INode** begin() { return array; }
+            const INode*const* begin() const { return array; }
+            INode** end() { return array+count; }
+            const INode*const* end() const { return array+count; }
+
+            inline INode*& operator[](size_t i) { assert(i<count); return array[i]; }
+            inline const INode* const& operator[](size_t i) const { assert(i<count); return array[i]; }
+        };
+        template <typename ...Contents>
+        static inline children_array_t createChildrenArray(Contents... children) 
+        { 
+            children_array_t a;
+            const INode* ch[]{ children... };
+            memcpy(a.array, ch, sizeof(ch));
+            a.count = sizeof...(children);
+            return a; 
+        }
+
+        using color_t = core::vector3df_SIMD;
+
+        explicit INode(E_SYMBOL s) : symbol(s) {}
+        virtual ~INode() = default;
+
+        children_array_t children;
+        E_SYMBOL symbol;
+        bool deinited = false;
+    };
+
+    struct CMaterialNode : public INode
+    {
+        CMaterialNode() : INode(ES_MATERIAL), thin(false)
+        {
+            opacity.source = EPS_CONSTANT;
+            opacity.value.constant = color_t(1.f);
+        }
+
+        SParameter<color_t> opacity;
+        bool thin;
+    };
+
+    struct CGeomModifierNode : public INode
+    {
+        enum E_TYPE
+        {
+            ET_DISPLACEMENT,
+            ET_HEIGHT,
+            ET_NORMAL,
+            ET_DERIVATIVE
+        };
+        enum E_SOURCE
+        {
+            ESRC_UV_FUNCTION,
+            ESRC_TEXTURE
+        };
+
+        CGeomModifierNode(E_TYPE t) : INode(ES_GEOM_MODIFIER), type(t) {}
+
+        E_TYPE type;
+        E_SOURCE source;
+        //TODO some union member for when source==ESRC_UV_FUNCTION, no idea what type
+        //in fact when we want to translate MDL function of (u,v) into this IR we could just create an image being a 2D plot of this function with some reasonable quantization (pixel dimensions)
+        //union {
+            STextureSource texture;
+        //};
+    };
+
+    struct CEmissionNode : INode
+    {
+        CEmissionNode() : INode(ES_EMISSION) {}
+
+        float intensity;
+    };
+
+    struct CBSDFCombinerNode : INode
+    {
+        enum E_TYPE
+        {
+            //mix of N BSDFs
+            ET_MIX,
+            //blend of 2 BSDFs weighted by constant or texture
+            ET_WEIGHT_BLEND,
+            //blend of 2 BSDFs weighted by fresnel
+            ET_FRESNEL_BLEND,
```

Comment:
what do we use that one for?

Reply:
fresnel - dielectric
weight - blend

Reply:
I dont get the explanation

Reply:
anyway.. when dielectric gets its dedicated IR node, it will not be used by anything in mitsuba frontend. But it is present in MDL under the name `df::fresnel_layer`. See http://mdlhandbook.com/mdl_handbook/index.html#combining_distribution_functions#fresnel-layering-in-a-material
`ET_WEIGHT_BLEND` is however used for encoding mitsuba's `blendbsdf` and `mixturebsdf`

Reply:
why cant the same node node encode all types of weighing/mixing? we just need to discern what the weights would be...

Reply:
rename `ET_FRESNEL_BLEND` to `ET_COAT_BLEND`

if you want something for supporting the broken way MDL layers materials with fresnel reserve the name `ET_LOL_MDL_SUX_BROKEN_FRESNEL_BLEND`

Reply:
> rename ET_FRESNEL_BLEND to ET_COAT_BLEND

i've created separate IR node for coating, so i can remove FRESNEL_BLEND just as well

Reply:
I think `ET_LOL_MDL_SUX_BROKEN_FRESNEL_BLEND` is a much better idea XD